File: publication/shader-based-astronomy/python/src/shader_astronomy/coords.py
"""S-entropy coordinate system.

Implements Definition 3.1 of the paper: the dimensionless mapping between
atmospheric thermodynamic state (T, P, rho, v, E) and texture-native
coordinates (S_k, S_t, S_e) in [0,1]^3.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

# Physical constants (SI)
K_B = 1.380649e-23  # Boltzmann
N_A = 6.02214076e23  # Avogadro
R_GAS = K_B * N_A  # gas constant
M_AIR = 28.9647e-3 / N_A  # mean molecular mass, kg

# Atmospheric reference ranges (SI). These are the [min, max] windows
# that map onto the [0, 1] coordinate range.
E_KIN_MIN, E_KIN_MAX = 1.0e-21, 2.0e-20  # J per molecule, 0 to 1000 K
TAU_MIN, TAU_MAX = 1.0e-10, 1.0e-8  # collision interval, s
E_TOT_MIN, E_TOT_MAX = 1.0e-21, 3.0e-20  # J per molecule


@dataclass(frozen=True)
class SEntropy:
    """Three-channel S-entropy coordinate at a voxel."""

    s_k: float  # kinetic / compositional
    s_t: float  # temporal / velocity
    s_e: float  # energy

    def as_array(self) -> NDArray[np.float64]:
        return np.array([self.s_k, self.s_t, self.s_e], dtype=np.float64)

    def __post_init__(self) -> None:
        for name, v in (("s_k", self.s_k), ("s_t", self.s_t), ("s_e", self.s_e)):
            if not 0.0 <= v <= 1.0:
                raise ValueError(f"{name}={v} outside [0,1]")


def physical_to_s(
    temperature_k: float,
    pressure_pa: float,
    velocity_m_s: float,
) -> SEntropy:
    """Map physical state to S-entropy coordinates (Eqs. 4-6 of paper).

    Parameters
    ----------
    temperature_k : absolute temperature, K
    pressure_pa : pressure, Pa
    velocity_m_s : wind speed magnitude, m/s
    """
    # Mean kinetic energy per molecule: (3/2) k_B T
    e_kin = 1.5 * K_B * temperature_k
    s_k = (e_kin - E_KIN_MIN) / (E_KIN_MAX - E_KIN_MIN)

    # Collision time: mean free path / thermal velocity.
    # Simplified: tau ~ k_B T / (P * sigma_collision * v_thermal)
    v_therm = np.sqrt(3.0 * K_B * temperature_k / M_AIR)
    sigma_coll = 4.0e-19  # m^2, typical air
    n_density = pressure_pa / (K_B * temperature_k)
    tau_coll = 1.0 / (n_density * sigma_coll * v_therm)
    s_t = (tau_coll - TAU_MIN) / (TAU_MAX - TAU_MIN)

    # Total energy per molecule (kinetic + translation from bulk wind)
    e_bulk = 0.5 * M_AIR * velocity_m_s**2
    e_tot = e_kin + e_bulk
    s_e = (e_tot - E_TOT_MIN) / (E_TOT_MAX - E_TOT_MIN)

    return SEntropy(
        s_k=float(np.clip(s_k, 0.0, 1.0)),
        s_t=float(np.clip(s_t, 0.0, 1.0)),
        s_e=float(np.clip(s_e, 0.0, 1.0)),
    )


def s_to_physical(s: SEntropy) -> tuple[float, float, float]:
    """Inverse map. Returns (T [K], P [Pa], |v| [m/s])."""
    e_kin = s.s_k * (E_KIN_MAX - E_KIN_MIN) + E_KIN_MIN
    temperature_k = e_kin / (1.5 * K_B)

    tau_coll = s.s_t * (TAU_MAX - TAU_MIN) + TAU_MIN
    v_therm = np.sqrt(3.0 * K_B * temperature_k / M_AIR)
    sigma_coll = 4.0e-19
    n_density = 1.0 / (tau_coll * sigma_coll * v_therm)
    pressure_pa = n_density * K_B * temperature_k

    e_tot = s.s_e * (E_TOT_MAX - E_TOT_MIN) + E_TOT_MIN
    e_bulk = max(e_tot - e_kin, 0.0)
    velocity_m_s = np.sqrt(2.0 * e_bulk / M_AIR)

    return float(temperature_k), float(pressure_pa), float(velocity_m_s)


def refractive_index(density_kg_m3: float) -> float:
    """Atmospheric refractive index from Lorentz-Lorenz (paper Eq. 8)."""
    rho_0 = 1.225  # kg/m^3 at sea level
    k_ref = 2.9e-4
    return 1.0 + k_ref * density_kg_m3 / rho_0


def atmospheric_density(altitude_m: float) -> float:
    """Four-mechanism density profile (paper Pass 0).

    rho(z) = sum_j rho_{0,j} * exp(-z/H_j) with four exponential layers.
    """
    rho_dry = 1.205 * np.exp(-altitude_m / 8500.0)
    rho_vapour = 0.013 * np.exp(-altitude_m / 2000.0)
    rho_trace = 0.001 * np.exp(-altitude_m / 5000.0)
    rho_aerosol = 1.5e-5 * np.exp(-altitude_m / 1500.0)
    return float(rho_dry + rho_vapour + rho_trace + rho_aerosol)

File: publication/shader-based-astronomy/rust/src/coords.rs
//! S-entropy coordinate system (paper Def. 3.1).
//!
//! Mirrors the Python reference in `shader_astronomy.coords` exactly.

use bytemuck::{Pod, Zeroable};

/// Boltzmann constant (SI).
pub const K_B: f32 = 1.380_649e-23;

/// Range bounds for S_k (Eq. 4).
pub const E_KIN_MIN: f32 = 1.0e-21;
/// Range bound.
pub const E_KIN_MAX: f32 = 2.0e-20;
/// Range bound.
pub const TAU_MIN: f32 = 1.0e-10;
/// Range bound.
pub const TAU_MAX: f32 = 1.0e-8;
/// Range bound.
pub const E_TOT_MIN: f32 = 1.0e-21;
/// Range bound.
pub const E_TOT_MAX: f32 = 3.0e-20;

/// Three-channel S-entropy coordinate stored per voxel.
#[repr(C)]
#[derive(Debug, Clone, Copy, Pod, Zeroable)]
pub struct SEntropy {
    /// Kinetic / compositional channel (maps temperature).
    pub s_k: f32,
    /// Temporal / velocity channel (maps collision interval).
    pub s_t: f32,
    /// Energy channel (maps total energy).
    pub s_e: f32,
    /// Refractive index (cached).
    pub n_ref: f32,
}

impl SEntropy {
    /// Construct from physical state.
    pub fn from_physical(temperature_k: f32, pressure_pa: f32, velocity_m_s: f32) -> Self {
        let e_kin = 1.5 * K_B * temperature_k;
        let s_k = ((e_kin - E_KIN_MIN) / (E_KIN_MAX - E_KIN_MIN)).clamp(0.0, 1.0);

        let m_air = 4.8096e-26_f32;
        let v_therm = (3.0 * K_B * temperature_k / m_air).sqrt();
        let sigma_coll = 4.0e-19_f32;
        let n_density = pressure_pa / (K_B * temperature_k);
        let tau = 1.0 / (n_density * sigma_coll * v_therm);
        let s_t = ((tau - TAU_MIN) / (TAU_MAX - TAU_MIN)).clamp(0.0, 1.0);

        let e_bulk = 0.5 * m_air * velocity_m_s * velocity_m_s;
        let e_tot = e_kin + e_bulk;
        let s_e = ((e_tot - E_TOT_MIN) / (E_TOT_MAX - E_TOT_MIN)).clamp(0.0, 1.0);

        let rho = (pressure_pa * m_air) / (K_B * temperature_k);
        let n_ref = 1.0 + 2.9e-4 * rho / 1.225;

        Self { s_k, s_t, s_e, n_ref }
    }
}

/// USSA-76-style atmospheric density at altitude.
pub fn atmospheric_density(altitude_m: f32) -> f32 {
    let rho_dry = 1.205 * (-altitude_m / 8500.0).exp();
    let rho_vapour = 0.013 * (-altitude_m / 2000.0).exp();
    let rho_trace = 0.001 * (-altitude_m / 5000.0).exp();
    let rho_aerosol = 1.5e-5 * (-altitude_m / 1500.0).exp();
    rho_dry + rho_vapour + rho_trace + rho_aerosol
}

/// Lorentz-Lorenz refractive index from mass density (Eq. 8).
pub fn refractive_index(density_kg_m3: f32) -> f32 {
    const RHO_0: f32 = 1.225;
    const K_REF: f32 = 2.9e-4;
    1.0 + K_REF * density_kg_m3 / RHO_0
}
